src/Control/Concurrent/CachedIO.hs

```diff
@@ -3,69 +3,76 @@ module Control.Concurrent.CachedIO (
+cachedIO
+    :: (MonadIO m, MonadIO t, MonadCatch t)
+    => NominalDiffTime -- ^ Number of seconds before refreshing cache
+    -> t a -- ^ action to cache
+    -> m (t a)
 cachedIO interval = cachedIOWith (secondsPassed interval)
 
 -- | Check if @starting time@ + @seconds@ is after @end time@
-secondsPassed :: NominalDiffTime  -- ^ Seconds
-               -> UTCTime         -- ^ Start time
-               -> UTCTime         -- ^ End time
-               -> Bool
+secondsPassed
+    :: NominalDiffTime -- ^ Seconds
+    -> UTCTime -- ^ Start time
+    -> UTCTime -- ^ End time
+    -> Bool
 secondsPassed interval start end = addUTCTime interval start > end
 
--- | Cache an IO action, The cache begins uninitialized.
+-- | Cache an action, The cache begins uninitialized.
 --
--- The outer IO is responsible for setting up the cache. Use the inner one to
+-- The outer 'MonadIO' is responsible for setting up the cache. Use the inner one to
 -- either get the cached value or refresh
-cachedIOWith :: (MonadIO m)
+cachedIOWith
+    :: (MonadIO m, MonadIO t, MonadCatch t)
     => (UTCTime -> UTCTime -> Bool) -- ^ Test function:
     --   If 'isCacheStillFresh' 'lastUpdated' 'now' returns 'True'
-    --   the cache is considered still fresh and returns the cached IO action
-    -> m a                          -- ^ action to cache
-    -> m (m a)
+    --   the cache is considered still fresh and returns the cached action
+    -> t a -- ^ action to cache
+    -> m (t a)
 cachedIOWith isCacheStillFresh io = do
-  initTime <- liftIO getCurrentTime
-  cachedT <- liftIO (atomically (newTVar (initTime, Left Uninitialized)))
+  cachedT <- liftIO (atomically (newTVar Uninitialized))
   return $ do
     now <- liftIO getCurrentTime
     join . liftIO . atomically $ do
       cached <- readTVar cachedT
       case cached of
+        (Fresh lastUpdated value)
         -- There's data in the cache and it's recent. Just return.
-        (lastUpdated, Right value) | isCacheStillFresh lastUpdated now ->
-          return (return value)
-        -- There's data in the cache, but it's stale. Update the cache timestamp
-        -- to prevent a second thread from also executing the IO action. The second
-        -- thread would get the stale data instead. Refresh the cache and return.
-        (_, Right value) -> do
-          writeTVar cachedT (now, Right value)
-          return $ refreshCache now cachedT
+          | isCacheStillFresh lastUpdated now -> return (return value)
+        -- There's data in the cache, but it's stale. Update the cache state
+        -- to prevent a second thread from also executing the action. The second
+        -- thread will get the stale data instead.
+          | otherwise -> do
+            writeTVar cachedT (Updating lastUpdated value)
+            return $ refreshCache (Fresh lastUpdated value) cachedT
+        -- Another thread is already updating the cache, just return the stale value
+        Updating _ value -> return (return value)
```

Comment:
Come to think of it I suppose `Updating UTCTime a` should simplify to `Updating a`